### PR TITLE
Update overlap.py to fix_factorial2

### PR DESCRIPTION
**Pull request by LudoRNLT**
Capture d'écran 2024-03-07 104652
Capture d'écran 2024-03-07 104853

@D-TheProgrammer and I, think that we solved the factorial error. Let us know if we have to do this but in a different way or use another syntax.

### CODE REVIEW

**Review by PaulWAyers — APPROVED** [review]
This looks good to me. I've tagged @FarnazH as she did more work on this, so I'd wait a couple days for her to glance at it before merging.

**Comment by D-TheProgrammer** [comment]
Hi @PaulWAyers  and @FarnazH , @LudoRNLT and I wanted to know if you have any updates regarding our fix and if it still suits you for the merge.

**Comment by PaulWAyers** [comment]
I think @FarnazH is busy. Perhaps @tovrstra or @marco-2023 has time to do a quick review.

**Comment by tovrstra** [comment]
@D-TheProgrammer: can you explain the error you were getting and that was fixed with this change? I just ran all unit tests without this change, and they pass (except for the problem fixed in #310).

File: iodata/overlap.py [hunk 1]
```diff
@@ -45,9 +45,14 @@ def factorial2(n, exact=False):
     """
     # Scipy  1.11.x returns an integer when n is an integer, but 1.10.x returns an array,
     # so np.array(n) is passed to make sure the output is always an array.
-    out = scipy.special.factorial2(np.array(n), exact=exact)
-    out[out <= 0] = 1.0
-    out[out <= -2] = 0.0
+    if exact == False : #Keep when exact=False
+        out = scipy.special.factorial2(np.array(n), exact=exact)
+    else :
+        if n is not int:
```
Comment (tovrstra) [inline iodata/overlap.py:51]: Please fix the type checking:
```suggestion
        if not isinstance(n, int):
```

File: iodata/overlap.py [hunk 2]
```diff
@@ -45,9 +45,14 @@ def factorial2(n, exact=False):
...
+            n = np.array([n]) # convert n to a np.array
+            for x in range(len(n)):
+                x = n[x]
+                out = scipy.special.factorial2(x, exact) #compute on each individual element x of the array n
```
Comment (tovrstra) [inline iodata/overlap.py:55]: Shouldn't this loop be dedented by four spaces?

File: iodata/overlap.py [hunk 3]
```diff
@@ -45,9 +45,14 @@ def factorial2(n, exact=False):
     """
     # Scipy  1.11.x returns an integer when n is an integer, but 1.10.x returns an array,
     # so np.array(n) is passed to make sure the output is always an array.
-    out = scipy.special.factorial2(np.array(n), exact=exact)
-    out[out <= 0] = 1.0
-    out[out <= -2] = 0.0
+    if exact == False : #Keep when exact=False
```
Comment (tovrstra) [inline iodata/overlap.py:48]:
```suggestion
    if not exact:
```

**Review by tovrstra — CHANGES_REQUESTED** [review]
I'm sorry for my initial confusion. I'd like to request a few improvements:

- Major: Please add unit tests for the cases you're fixing. The issue reported in #308 is not covered by our unit tests at the moment. Unit tests for the following cases (once with `exact=True` and once with `exact=False`) would be useful:
    - integer arguments `0`, `1`, `2` and `3`
    - float arguments `0.0`, `1.0`, `2.0` and `3.0`
    - integer array argument `np.array([0, 1, 2, 3])`
    - float array argument `np.array([0.0, 1.0, 2.0, 3.0])`
- Minor: Please avoid comments at the end of a line. Put them on the line before instead. Also, don't add comments with little added value, such as translating a line of Python to a human-readable sentence. Comments are more useful when they explain the intent of the code, typically of a block consisting of multiple lines, which is different from a literal human translation of the code.

Thanks!

**Comment by tovrstra** [comment]
I'm closing this one because a more recent pull request was made: #319

**Pull request closed without merging**